packages/react-router-redux/modules/ConnectedRouter.js

```diff
@@ -15,22 +15,73 @@ class ConnectedRouter extends Component {
   }
 
   handleLocationChange = location => {
-    this.store.dispatch({
-      type: LOCATION_CHANGE,
-      payload: location
-    })
+    /*
+     * Update the store in 2 cases:
+     * 1) Not syncing history to match store (normal flow from history to store).
+     * 2) DevTools resets state to empty location (need to change to an initial
+     * location).
+    */
+    if (!this.isSyncingHistory || !this.store.getState().router.location.pathname) {
+      this.store.dispatch({
+        type: LOCATION_CHANGE,
+        payload: location
+      })
+    }
+    // Stop syncing the history.
+    this.isSyncingHistory = false
+  }
+
+  handleStoreChange = () => {
+    // Extract store's location.
+    const storeLocation = this.store.getState().router.location
+    const {
+      pathname: storePathname,
+      search: storeSearch,
+      hash: storeHash,
+    } = storeLocation
+    // Extract history's location.
+    const {
+      pathname: historyPathname,
+      search: historySearch,
+      hash: historyHash,
+    } = this.props.history.location
+
+    // In normal situation, the store receives location state from the history.
+    // When the store is updated, the location in the store will be in sync with
+    // the location in the history. However, if the location in the store is
+    // updated before the history, they will be different. We might be time
+    // travelling, replacing root reducer, importing state, etc. So, we need to
+    // update the history to sync with the store.
+    if (historyPathname !== storePathname
+      || historySearch !== storeSearch
+      || historyHash !== storeHash) {
+      // start syncing the history
+      this.isSyncingHistory = true
+
+      if (!storePathname) {
+        // Use initial location if the store is reset.
+        // Store will be updated to initial location in handleLocationChange.
+        this.props.history.push(this.initialLocation)
+      } else {
+        this.props.history.push(storeLocation)
```

Comment:
This is going to break the browser's history. You cannot just `push` the location onto the history stack because this will create a new entry in the history. This means you're always going forward in the history, even if the intent was to go back.

What you really need to do here is `history.go(n)` where `n` is the difference in the index of the current location in the history and the index of the `storeLocation`, or the location you want to get to. But as much as I've tried to think about it over the years, I can't figure out a reliable way to get that `n`.